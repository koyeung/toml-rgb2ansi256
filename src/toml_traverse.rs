//! Traverse TOML document.
use toml_edit::{Item, Table, Value};

/// Apply function on `Value` node while traverse from input `Item`.
pub fn visit_item<F: Fn(&mut Value) + Copy>(item: &mut Item, f: F) {
    match item {
        Item::Value(value) => visit_value(value, f),
        Item::Table(table) => visit_table(table, f),
        Item::ArrayOfTables(array_of_tables) => {
            array_of_tables.iter_mut().for_each(|t| visit_table(t, f))
        }
        Item::None => (),
    }
}

fn visit_table<F: Fn(&mut Value) + Copy>(table: &mut Table, f: F) {
    for (_, item) in table.iter_mut() {
        visit_item(item, f);
    }
}

fn visit_value<F: Fn(&mut Value) + Copy>(value: &mut Value, f: F) {
    match value {
        Value::InlineTable(inline_table) => {
            inline_table.iter_mut().for_each(|(_, v)| visit_value(v, f))
        }
        Value::Array(array) => array.iter_mut().for_each(|v| visit_value(v, f)),
        _ => f(value),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use toml_edit::DocumentMut;

    #[test]
    fn test_modify_values() {
        const INPUT: &str = r##"
        # this is comment

        a = "a"
        "a.b" = "a.b"          # comment on keyword.directive
        c = { d = "c.d" }

        [p]
        "a" = "p.a"  # some comment
        "b" = "p.b"

        [[q]]
        a = "a1"

        [[q]]
        a = "a2"
        "##;

        let mut doc = INPUT.parse::<DocumentMut>().expect("doc should be in toml");

        visit_item(doc.as_item_mut(), |v| {
            if let Value::String(s) = v {
                let old_str_value = s.value();

                let decor = s.decor();
                let prefix = decor.prefix().unwrap();
                let suffix = decor.suffix().unwrap();

                let new_v = Value::from(old_str_value.to_uppercase());
                let decorated_new_v = new_v.decorated(prefix.clone(), suffix.clone());

                *v = decorated_new_v;
            }
        });

        insta::assert_toml_snapshot!(doc.to_string());
    }
}
