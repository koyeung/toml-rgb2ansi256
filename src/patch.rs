//! Patch RGB value to ANSI-256 value in TOML document.

use toml_edit::{Document, Value};

use crate::toml_traverse::visit_item;

/// Transform value node of RGB hex values to ANSI-256 color index.
pub fn patch_doc(doc: &str) -> String {
    let mut doc = doc.parse::<Document>().expect("should be proper toml");

    visit_item(doc.as_item_mut(), |v| {
        if let Value::String(s) = v {
            if let Ok(Rgb { r, g, b }) = string_to_rgb(s.value()) {
                let ansi_256 = rgb2ansi256::rgb_to_ansi256(r, g, b);

                // extract old decoration
                let decor = s.decor();
                let prefix = decor.prefix().unwrap();
                let suffix = decor.suffix().unwrap();

                // wrap ansi_256 with decoration
                let new_value =
                    Value::from(ansi_256.to_string()).decorated(prefix.clone(), suffix.clone());

                *v = new_value;
            }
        }
    });

    doc.to_string()
}

struct Rgb {
    r: u8,
    g: u8,
    b: u8,
}

fn string_to_rgb(s: &str) -> Result<Rgb, String> {
    if s.starts_with('#') && s.len() >= 7 {
        if let (Ok(r), Ok(g), Ok(b)) = (
            u8::from_str_radix(&s[1..3], 16),
            u8::from_str_radix(&s[3..5], 16),
            u8::from_str_radix(&s[5..7], 16),
        ) {
            Ok(Rgb { r, g, b })
        } else {
            Err(format!("bad rgb hexcode: {}", s))
        }
    } else {
        Err(format!("{} is not a rgb str", s))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_rgb_value_to_ansi256() {
        const INPUT: &str = r##"
        a = "unchanged"
        b = "#00ffaf"
        "##;

        insta::assert_toml_snapshot!(patch_doc(INPUT));
    }
}
