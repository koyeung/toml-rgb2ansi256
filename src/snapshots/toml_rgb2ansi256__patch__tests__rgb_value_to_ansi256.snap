---
source: src/patch.rs
expression: patch_doc(INPUT)
---
'''

        a = "unchanged"
        b = "49"
        '''
